<div style="background-color: #86CBBB; 1px; height:3px " ></div>

# The end: Optional integrative exercise


In your research group you're analyzing the proteome (the set of proteins expressed by an organism or cell) of different species.

Integrating different commands that you have learn during this practical, answer the following questions:

1. Once you have downloaded the data (you cannot use `wget` with this link, so download it from the browser), unzip the file with the proper function according to the type of compression.
2. Indicate the basic properties of this data: how many folders and files have you unzipped? How is the size of the files and folders? How many lines do they have? etc.

Move the entire folder to the desktop, into a folder called <code>IntegrativeExerciseP1</code>. Answer the following biological questions:

1. How many proteins does the human proteome have? And the mouse? (The human proteome is stored in the file <code>Homo_sapiens.GRCh38.pep.all.fa</code>and the mouse proteome in <code>Mus_musculus.GRCm38.pep.all.fa</code> and has the structure of a regular fasta file)
2. Find, for all the files that contain the longest protein (the files that have the extension `.liso`), how many sequences they have contained
3. List all protein names without the comment (the comment always starts after the first space in the sequence identifier of the fasta file)
4. How many genes are contained in the human genome?
5. How many genes have a single isoform?
6. How many genes have more than one isoform?
7. How many domains are there in the human proteome? How many are different? (The domains are in the file `Homo_sapiens.GRCh38.pep.all.fa.liso.dom` and always start with the identifier PF. Since the file is separated with multiple spaces, use the command `sed -r 's/ +/ /g '` to remove them by a single space first)
8. How many times it is found the 19th most frequent domain? Which domain it is?


<div style="background-color: #86CBBB; 1px; height:3px " ></div>

#### We have to change the working directory to the one which contains the zip file and unzip it

In [8]:
%%bash
cd ../../P1_TAB

unzip PFAM_seqs.zip

Archive:  PFAM_seqs.zip
   creating: pfam/
  inflating: pfam/Gallus_gallus.Galgal4.pep.all.fa.liso.dom  
  inflating: pfam/Homo_sapiens.GRCh38.pep.all.fa.liso.dom  
  inflating: pfam/Mus_musculus.GRCm38.pep.all.fa.liso.dom  
   creating: seqs/
  inflating: seqs/Gallus_gallus.Galgal4.pep.all.fa.liso  
  inflating: seqs/Homo_sapiens.GRCh38.pep.all.fa  
  inflating: seqs/Homo_sapiens.GRCh38.pep.all.fa.liso  
  inflating: seqs/Mus_musculus.GRCm38.pep.all.fa  
  inflating: seqs/Mus_musculus.GRCm38.pep.all.fa.liso  


<div style="background-color: #86CBBB; 1px; height:3px " ></div>

#### Indicate the basic properties of this data: how many folders and files have you unzipped? How is the size of the files and folders? How many lines do they have? etc.

In [17]:
%%bash

cd ../../P1_TAB

ls

animals.txt
book_author.txt
book_title.txt
book_translate.txt
colors.txt
Dorian_Gray.txt
pfam
PFAM_seqs.zip
seqs


When we unzip the downloaded file, we obtain **two different folders**:
- pfam
- seqs

<div style="background-color: #86CBBB; 1px; height:3px " ></div>

In [15]:
%%bash

cd ../../P1_TAB/pfam

ls

Gallus_gallus.Galgal4.pep.all.fa.liso.dom
Homo_sapiens.GRCh38.pep.all.fa.liso.dom
Mus_musculus.GRCm38.pep.all.fa.liso.dom


The "pfam" folder contains **three files** with information from:
- _Gallus gallus_
- _Homo Sapiens_
- _Mus musculus_

<div style="background-color: #86CBBB; 1px; height:3px " ></div>

In [16]:
%%bash

cd ../../P1_TAB/seqs

ls

Gallus_gallus.Galgal4.pep.all.fa.liso
Homo_sapiens.GRCh38.pep.all.fa
Homo_sapiens.GRCh38.pep.all.fa.liso
Mus_musculus.GRCm38.pep.all.fa
Mus_musculus.GRCm38.pep.all.fa.liso


The "seqs" folder contains **five files** with information from:
- _Gallus gallus_
- _Homo Sapiens_ x2
- _Mus musculus_ x2

<div style="background-color: #86CBBB; 1px; height:3px " ></div>

In [29]:
%%bash

cd ../../P1_TAB/

#pfam folder size
du -h pfam

#pfam files size
cd pfam

du -h Gallus_gallus.Galgal4.pep.all.fa.liso.dom

du -h Homo_sapiens.GRCh38.pep.all.fa.liso.dom

du -h Mus_musculus.GRCm38.pep.all.fa.liso.dom

#seqs folder size
cd ..

du -h seqs

#seqs files size
cd seqs

du -h Gallus_gallus.Galgal4.pep.all.fa.liso

du -h Homo_sapiens.GRCh38.pep.all.fa

du -h Homo_sapiens.GRCh38.pep.all.fa.liso

du -h Mus_musculus.GRCm38.pep.all.fa

du -h Mus_musculus.GRCm38.pep.all.fa.liso

16M	pfam
4,0M	Gallus_gallus.Galgal4.pep.all.fa.liso.dom
6,1M	Homo_sapiens.GRCh38.pep.all.fa.liso.dom
5,8M	Mus_musculus.GRCm38.pep.all.fa.liso.dom
156M	seqs
13M	Gallus_gallus.Galgal4.pep.all.fa.liso
66M	Homo_sapiens.GRCh38.pep.all.fa
19M	Homo_sapiens.GRCh38.pep.all.fa.liso
42M	Mus_musculus.GRCm38.pep.all.fa
18M	Mus_musculus.GRCm38.pep.all.fa.liso


The command **du -h**, shows the size of the folder/file, followed by the name of the folder or file.


<div style="background-color: #86CBBB; 1px; height:3px " ></div>

In [30]:
%%bash

#pfam files lines
cd ../../P1_TAB/pfam

wc -l Gallus_gallus.Galgal4.pep.all.fa.liso.dom

wc -l Homo_sapiens.GRCh38.pep.all.fa.liso.dom

wc -l Mus_musculus.GRCm38.pep.all.fa.liso.dom

#seqs files lines
cd ../seqs


wc -l Gallus_gallus.Galgal4.pep.all.fa.liso

wc -l Homo_sapiens.GRCh38.pep.all.fa

wc -l Homo_sapiens.GRCh38.pep.all.fa.liso

wc -l Mus_musculus.GRCm38.pep.all.fa

wc -l Mus_musculus.GRCm38.pep.all.fa.liso

29955 Gallus_gallus.Galgal4.pep.all.fa.liso.dom
47173 Homo_sapiens.GRCh38.pep.all.fa.liso.dom
43355 Mus_musculus.GRCm38.pep.all.fa.liso.dom
146255 Gallus_gallus.Galgal4.pep.all.fa.liso
796078 Homo_sapiens.GRCh38.pep.all.fa
215686 Homo_sapiens.GRCh38.pep.all.fa.liso
511577 Mus_musculus.GRCm38.pep.all.fa
205169 Mus_musculus.GRCm38.pep.all.fa.liso


The command **wc -l**, shows the lines of the file, followed by the file's name.
<div style="background-color: #86CBBB; 1px; height:3px " ></div>

In [44]:
%%bash

cd ../../Escritorio
mkdir IntegrativeExerciseP1

In [51]:
%%bash

cd ../../P1_TAB
mv pfam ../Escritorio/IntegrativeExerciseP1/pfam

mv seqs ../Escritorio/IntegrativeExerciseP1/seqs

cd ../Escritorio/IntegrativeExerciseP1
ls

pfam
seqs


<div style="background-color: #86CBBB; 1px; height:3px " ></div>

#### 1. How many proteins does the human proteome have? And the mouse?

In [1]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

grep -c ">" Homo_sapiens.GRCh38.pep.all.fa 
grep -c ">" Mus_musculus.GRCm38.pep.all.fa

102915
58668


- The human proteome have **102915 proteins**.
- The mouse proteome have **58668 proteins**.

*To find the total number of proteins, as we know that it is a fasta file, we know that they start with the symbol ">", so we count how many times this symbol appears.*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 2. Find, for all the files that contain the longest protein (the files that have the extension .liso), how many sequences they have contained.

In [2]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

grep -c ">" Gallus_gallus.Galgal4.pep.all.fa.liso
grep -c ">" Homo_sapiens.GRCh38.pep.all.fa.liso
grep -c ">" Mus_musculus.GRCm38.pep.all.fa.liso

15508
22964
22769


- The *Gallus gallus* file contains **15508 sequences**.
- The *Homo sapiens* file contains **22964 sequences**.
- The *Mus musculus* file contains **22769 sequences**.


*In this case we look at the folders containing the longest protein encoded by each gene. To do this, we also count how many times the character ">" appears.*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 3. List all protein names without the comment.

In [1]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

cut -f1 -d" "  Homo_sapiens.GRCh38.pep.all.fa | grep ">" | cut -f2 -d">" | head -n 20

ENSP00000452494.1
ENSP00000488240.1
ENSP00000487941.1
ENSP00000451515.1
ENSP00000451042.1
ENSP00000488261.1
ENSP00000487812.1
ENSP00000475053.2
ENSP00000417751.1
ENSP00000488475.1
ENSP00000418639.1
ENSP00000487775.1
ENSP00000488083.1
ENSP00000488720.1
ENSP00000487789.1
ENSP00000488168.1
ENSP00000419139.1
ENSP00000488589.1
ENSP00000487937.1
ENSP00000419564.1


In [2]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

cut -f1 -d" "  Mus_musculus.GRCm38.pep.all.fa | grep ">" | cut -f2 -d">" | head -n 20

ENSMUSP00000142546.1
ENSMUSP00000142955.1
ENSMUSP00000141764.1
ENSMUSP00000141312.1
ENSMUSP00000142153.1
ENSMUSP00000141970.1
ENSMUSP00000142162.1
ENSMUSP00000141399.1
ENSMUSP00000141414.1
ENSMUSP00000141374.1
ENSMUSP00000141376.1
ENSMUSP00000141852.1
ENSMUSP00000141615.1
ENSMUSP00000141202.1
ENSMUSP00000141703.1
ENSMUSP00000141697.1
ENSMUSP00000142226.1
ENSMUSP00000142199.1
ENSMUSP00000141415.1
ENSMUSP00000142229.1


*First, we "cut" the lines according to the space separations, and field 1 corresponds to the protein identifier (cut -f1 -d" "), with which we are kept. Then, following the same procedure, we have "cut" the symbol ">", leaving us with the second field, which corresponds to the protein identifier without >.*

**NOTE:** I use the "head -n 20" command to show a shorter protein list (less file size), but without it, you can see all the proteins. 

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 4. How many genes are contained in the human genome?

In [13]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

cat Homo_sapiens.GRCh38.pep.all.fa| tr ' ' '\n' | grep "gene:E" | sort | uniq | wc -l

22964


- The *Homo sapiens* file contains **22964 genes** contained in the human genome.


*To do this, we divided the file by elements (each element to a line), and then we kept those lines that contained "gene:E" (to obtain the genes), we sorted, and then with "uniq" we kept only one repetition of each gene (since it is possible that several proteins come from the same gene) and finally we counted how many lines resulted.*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 5. How many genes have a single isoform?

In [14]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

cat Homo_sapiens.GRCh38.pep.all.fa| tr ' ' '\n' | grep "gene:E" | sort | uniq -u| wc -l

5730


- The *Homo sapiens* file contains **5730 genes** that have a single isoform.


*In this case we have to find the genes that have only one associated protein. To do so, we follow the same procedure as in the previous exercise, but in this case we use "uniq -u" because it allows us to obtain those lines that are not repeated (i.e. that only appear once), which correspond to genes that have only one isoform.*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 6. How many genes have more than one isoform?

In [15]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/seqs

cat Homo_sapiens.GRCh38.pep.all.fa| tr ' ' '\n' | grep "gene:E" | sort | uniq -d| wc -l

17234


- The *Homo sapiens* file contains **17234 genes** that have more than one isoform.


*In this case we have to make the opposite approach to the previous exercise, that is, to obtain those genes that have more than one associated protein (therefore they have more than one isoform). To do so, the procedure is the same but instead of using "uniq -u", we use "uniq -d", since it allows us to obtain the lines that are repeated (therefore the genes that appear more than once in the protein file).*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 7. How many domains are there in the human proteome? How many are different?

In [24]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/pfam

cat Homo_sapiens.GRCh38.pep.all.fa.liso.dom | sed "s/  */ /g"| cut -d" " -f6| grep "PF"| wc -l
cat Homo_sapiens.GRCh38.pep.all.fa.liso.dom | sed "s/  */ /g"| cut -d" " -f6| grep "PF"| uniq| wc -l

47144
34949


- The *Homo sapiens* file contains **47144 domains**, of which **34949** are different.


*Taking into account that the column containing the domain identifiers is the sixth one, we separated by spaces using "cut", and we kept the field 6. Then, we used "grep "PF"" to ensure that only the domain identifiers were taken into account, and finally we counted the resulting lines.*

*In the second section, to know how many are different, the only change is that we have to do "uniq", to obtain only one representation of each domain, so we can count how many are different.*

<div style="background-color: #86CBBB; 1px; height:3px " ></div>


#### 8. How many times it is found the 19th most frequent domain? Which domain it is?

In [25]:
%%bash

cd ../../Escritorio/IntegrativeExerciseP1/pfam

cat Homo_sapiens.GRCh38.pep.all.fa.liso.dom | sed "s/  */ /g"| cut -d" " -f6| grep "PF"| sort| uniq -c| sort -nr| head -n 19| tail -n 1

    293 PF00001.19


- The 19th most frequent domain (**PF00001.19**) is found **293 times**.


*First we have filtered also by the field containing the domain identifiers, then we have made a sort, to later count how many times each one appears (with the uniq -c). Afterwards, the sort -nr orders numerically and decreasingly the result, so that with the head and the tail we can keep the 19th result, which corresponds to the 19th most frequent domain.*